frontend/src/components/UserSearch.vue

```diff
@@ -334,21 +336,43 @@ export default {
           .find(client => this.clientId === client.id).name;
       return role;
     },
+    // Return only the users having the selected roles. Also add role information to the users.
     filterUsersByRole: function(searchResults, maxSearch) {
-      let roleRequests = this.selectedRoles.map(
+      let usersInRoleRequests = this.selectedRoles.map(
         clientRole => ClientsRepository
           .getUsersInRole(clientRole.clientId, clientRole.name, maxSearch)
+          .then(function (result) {
+            // The result doesn't include the role, so we include a reference to it here to use later.
+            return {'role': clientRole.name, 'result': result}
+          })
           .catch(error => {
             this.handleError(`Search failed for role ${clientRole.name}`, error);
           })
       );
-      return Promise.all(roleRequests)
-        .then(responses => responses.flatMap(
-          response => response.data.map(user => user.id)
-        ))
-        .then(userIds => searchResults.filter(
-          user => userIds.includes(user.id)
-        ));
+
+      return Promise.all(usersInRoleRequests)
+        .then(function (responses) {
+          let userIdsAndRoles = responses.flatMap(
+              response => response.result.data.map(function (user) {
+                return {'id': user.id, 'role': response.role}
+              })
+          );
+          for (let searchResult of searchResults) {
+            for (let user of userIdsAndRoles) {
+              if (searchResult.id == user.id) {
+                if (searchResult.role) {
+                  searchResult.role = searchResult.role + ", " + user.role;
+                } else {
+                  searchResult.role = user.role;
+                }
+              }
+            }
+          }
+          let userIds = userIdsAndRoles.map(userIdAndRole => userIdAndRole.id);
+          return searchResults.filter(
+              user => userIds.includes(user.id)
+          );
+        });
```

Comment:
I had a hard time writing this code because Promises and stream (map, flatMap, filter) are hard to wrap my head around, but I think the code is actually fairly clean and to the point.